annotation_pipeline/utils.py

```diff
@@ -91,53 +130,47 @@ def get_pixel_indices(coordinates):
     return pixel_indices
 
 
-def append_pywren_stats(futures, memory_mb, cloud_objects_n=0):
-    if type(futures) != list:
-        futures = [futures]
+def serialise_to_file(obj, path):
+    with open(path, 'wb') as file:
+        file.write(pa.serialize(obj).to_buffer())
 
-    def calc_cost(runtimes, memory_gb):
-        unit_price_in_dollars = 0.000017
-        return sum([unit_price_in_dollars * memory_gb * runtime for runtime in runtimes])
 
-    actions_num = len(futures)
-    func_name = futures[0].function_name
-    runtimes = [future.stats['exec_time'] for future in futures]
-    cost = calc_cost(runtimes, memory_mb / 1024)
-    headers = ['Function', 'Actions', 'Memory', 'AvgRuntime', 'Cost', 'CloudObjects']
-    content = [func_name, actions_num, memory_mb, np.average(runtimes), cost, cloud_objects_n]
+def deserialise_from_file(path):
+    with open(path, 'rb') as file:
+        data = pa.deserialize(file.read())
+    return data
 
-    Path('logs').mkdir(exist_ok=True)
-    if not Path(STATUS_PATH).exists():
-        with open(STATUS_PATH, 'w') as csvfile:
-            writer = csv.DictWriter(csvfile, fieldnames=headers)
-            writer.writeheader()
 
-    with open(STATUS_PATH, 'a') as csvfile:
-        writer = csv.DictWriter(csvfile, fieldnames=headers)
-        writer.writerow(dict(zip(headers, content)))
+def serialise(obj):
+    try:
+        return BytesIO(pa.serialize(obj).to_buffer())
+    except pa.lib.SerializationCallbackError:
+        return BytesIO(pickle.dumps(obj))
 
 
-def get_pywren_stats(log_path=STATUS_PATH):
-    stats = pd.read_csv(log_path)
-    print('Total PyWren cost: {:.3f} $'.format(stats['Cost'].sum()))
-    return stats
+def deserialise(serialised_obj):
+    data = serialised_obj.read()
```

Comment:
@LachlanStuart I dont get your notes about `read_object_with_retry` / `read_cloud_object_with_retry`. by reviewing its code again I can clearly see that if `stream_reader(data_stream)` or `data_stream.read()` fails, then it will go into "except" and then retrying by the for loop.

Reply:
@LachlanStuart I see, can you please elaborate about the benefits of the temp buffer incomparison to the logic that I implemented? also, are you familiar with a way to implement it with pyarrow?

Reply:
The benefit is just that it has a lower peak memory usage during deserialization. Deserializing a 100mb dataframe needs 200mb when using a temp buffer (100mb for the buffer + 100mb for the output object), or probably somewhere around 110mb for streaming (some small, fixed amount for a buffer + 100mb for the output object). 

I don't think this is a regression. Even though we were previously trying to use msgpack for streaming, I looked inside the Pandas code and internally it would just read the whole stream into a temporary buffer. Not sure about the non-Pandas msgpack though. However, looking at the pyarrow code it does look like streaming deserialization may be possible. Also, as you previously found, it's not possible to do streaming serialization to COS without using the smart-open library.

This doesn't need to be resolved. I was just surprised to see it after that Python upgrade. The request for changes was just for the 2 other small things.